crates/starknet-types-core/src/u256/mod.rs

```diff
@@ -0,0 +1,196 @@
+            FromStrError::Parse(e) => {
+                // Avoid using format as it requires `alloc`
+                core::fmt::Display::fmt("invalid string: ", f)?;
+                core::fmt::Display::fmt(e, f)
+            }
+            FromStrError::Invalid => core::fmt::Display::fmt("invalid characters", f),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for FromStrError {}
+
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct U256 {
+    high: u128,
+    low: u128,
+}
+
+impl U256 {
+    pub fn high(&self) -> u128 {
+        self.high
+    }
+
+    pub fn low(&self) -> u128 {
+        self.low
+    }
+
+    pub fn from_parts(high: u128, low: u128) -> Self {
+        Self { low, high }
+    }
+
+    pub fn try_from_felt_parts(high: Felt, low: Felt) -> Result<Self, PrimitiveFromFeltError> {
+        Ok(Self {
+            high: high.try_into()?,
+            low: low.try_into()?,
+        })
+    }
+
+    pub fn try_from_dec_str_parts(high: &str, low: &str) -> Result<Self, <u128 as FromStr>::Err> {
+        Ok(Self {
+            high: high.parse()?,
+            low: low.parse()?,
+        })
+    }
+
+    pub fn try_from_hex_str_parts(high: &str, low: &str) -> Result<Self, <u128 as FromStr>::Err> {
+        Ok(Self {
+            high: u128::from_str_radix(high, 16)?,
+            low: u128::from_str_radix(low, 16)?,
+        })
+    }
+
+    pub fn from_hex_str(hex_str: &str) -> Result<Self, FromStrError> {
+        let string_without_prefix = if hex_str.starts_with("0x") || hex_str.starts_with("0X") {
+            &hex_str[2..]
+        } else {
+            hex_str
+        };
+
+        let string_without_zero_padding = string_without_prefix.trim_start_matches('0');
+
+        let (high, low) = if string_without_zero_padding.is_empty() {
+            if string_without_zero_padding.len() == string_without_prefix.len() {
+                return Err(FromStrError::Invalid);
+            } else {
+                (0, 0)
+            }
+        } else if string_without_zero_padding.len() > 64 {
+            return Err(FromStrError::ValueTooBig);
+        } else if string_without_zero_padding.len() > 32 {
+            let delimiter_index = string_without_zero_padding.len() - 32;
+            (
+                u128::from_str_radix(&string_without_zero_padding[0..delimiter_index], 16)
+                    .map_err(FromStrError::Parse)?,
+                u128::from_str_radix(&string_without_zero_padding[delimiter_index..], 16)
+                    .map_err(FromStrError::Parse)?,
+            )
+        } else {
+            (
+                0,
+                u128::from_str_radix(string_without_zero_padding, 16)
+                    .map_err(FromStrError::Parse)?,
+            )
+        };
+
+        Ok(U256 { high, low })
+    }
+
+    pub fn from_dec_str(dec_str: &str) -> Result<Self, FromStrError> {
```

Comment:
why don't you use bigint ? it would be way less complex no ?

Reply:
i wouldn't call it straightforward 😅 Either add comments explaining what's happening or use the bigint crate so it's easier to understand

Reply:
I can add comments sure!